### PR TITLE
Finetuning on concatenated datasets

### PR DESCRIPTION
**Pull request by ryspark**
Finetuning on multiple datasets in memory at once.

- [X]  Collect test metrics at the end of a model test run
- [X]  Metrics, logging, checkpointing across different dataset heads
- [X]  Training/evaluation across different datasets
- [X]  Alternating batches between datasets
- [X] Fast distributed data loader that works across multiple ranks/processes
- [ ]  Mixed-dataset in a single batch
- [ ]  Multiple tasks within a single dataset

Currently only allows one task per dataset - this is the highest priority next fix, since some datasets (ex: v2_pastis) have ambiguous task names like detect or classify that collide with other datasets.

Occasionally there are NCCL hangs at the end of validation loops, this has been hard to reproduce/debug in isolation so unclear if this is due to multi-dataset training or a larger issue.

Also adds support for [ALPA](https://arxiv.org/html/2503.11335v2), but this isn't recommended as the performance gains during backward are quite small. Unclear why this is, maybe something to do with switching off gradients at such a finegrained parameter level?

### CODE REVIEW

**Review by favyen2 — APPROVED** [review]
This implementation is great! Having MultiDatasetDataModule create one data loader over a MultiDataset of ModelDatasets makes sense to me. Left a few minor comments.
